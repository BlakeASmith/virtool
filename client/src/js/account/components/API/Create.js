import CX from "classnames";
import React from "react";
import { push } from "react-router-redux";
import { connect } from "react-redux";
import { CopyToClipboard } from "react-copy-to-clipboard";
import { mapValues } from "lodash";
import { Col, Modal, Row } from "react-bootstrap";

import APIPermissions from "./Permissions";
import { Button, Icon, Input, Flex, FlexItem } from "../../../base/index";
import { routerLocationHasState } from "../../../utils";

const getInitialState = (props) => ({
    name: "",
    permissions: mapValues(props.permissions, () => false),
    submitted: false,
    copied: false,
    key: ""
});

export class CreateAPIKey extends React.Component {

    constructor (props) {
        super(props);
        this.state = getInitialState(props);
    }

    modalExited = () => {
        this.setState(getInitialState(this.props));
    };

    handleSubmit = (e) => {
        e.preventDefault();

        this.setState({submitted: true}, () => {
            this.props.onCreate(this.state.name, this.state.permissions, (key) => this.setState({key}));
        });
    };

    handlePermissionChange = (key, value) => {
        this.setState({permissions: {...this.state.permissions, [key]: value}});
    };

    render () {
        let content;

        if (this.state.key) {
            content = (
                <Modal.Body className="text-center">
                    <Row>
                        <Col xs={12}>
                            <strong className="text-success">Here is your key.</strong>
                        </Col>
                    </Row>

                    <small>
                        Make note of it now. For security purposes, it will not be shown again.
                    </small>

                    <Row style={{marginTop: "10px", marginBottom: "5px"}}>
                        <Col xs={12} md={8} mdOffset={2}>
                            <Flex alignItems="stretch" alignContent="stretch">
                                <FlexItem grow={1}>
                                    <Input
                                        style={{marginBottom: 0}}
                                        formGroupStyle={{marginBottom: 0}}
                                        className="text-center"
                                        value={this.state.key} onChange={() => {}}
                                    />
                                </FlexItem>
                                <CopyToClipboard
                                    text={this.state.key}
                                    onCopy={() => this.setState({copied: true})}
                                >
                                    <Button icon="paste" bsStyle="primary" />
                                </CopyToClipboard>
                            </Flex>
                        </Col>
                    </Row>

                    <small className={CX("text-primary", {invisible: !this.state.copied})}>
                        <Icon name="checkmark" /> Copied
                    </small>
                </Modal.Body>
            );
        } else {
            content = (
                <form onSubmit={this.handleSubmit}>
                    <Modal.Body>
                        <Input
                            label="Name"
                            value={this.state.name}
                            onChange={(e) => this.setState({name: e.target.value})}
                        />

                        <label>Permissions</label>

                        <APIPermissions
                            userPermissions={this.props.permissions}
                            keyPermissions={this.state.permissions}
                            onChange={this.handlePermissionChange}
                        />
                    </Modal.Body>

                    <Modal.Footer>
                        <Button type="submit" icon="floppy" bsStyle="primary">
                            Save
                        </Button>
                    </Modal.Footer>
                </form>
            );
        }

        return (
            <Modal show={this.props.show} onHide={this.props.onHide} onExited={this.modalExited}>
                <Modal.Header onHide={this.props.onHide} closeButton>
                    Create API Key
                </Modal.Header>

                {content}
            </Modal>
        );
    }
}

const mapStateToProps = (state) => ({
    show: routerLocationHasState(state, "createAPIKey"),
    permissions: state.account.permissions
});

const mapDispatchToProps = (dispatch) => ({

    onHide: () => {
        dispatch(push({...window.location, state: {createAPIKey: false}}));
    }

});

export default connect(mapStateToProps, mapDispatchToProps)(CreateAPIKey);
