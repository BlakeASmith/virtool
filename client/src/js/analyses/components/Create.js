import React from "react";
import { push } from "connected-react-router";
import { connect } from "react-redux";
import { get, map, find, forEach, xorBy } from "lodash-es";
import { Modal, ListGroup, Col, Label } from "react-bootstrap";
import { AlgorithmSelect, Button, ListGroupItem, NoneFound, Checkbox } from "../../base/index";
import { getTaskDisplayName, routerLocationHasState } from "../../utils/utils";
import { analyze } from "../actions";

export const IndexSelect = ({ indexes, onSelect, selected, error }) => {
    const errorStyle = error.length ? { border: "1px solid #d44b40" } : { border: "1px solid transparent" };

    return (
        <div style={{ marginBottom: "16px" }}>
            <label className="control-label">References</label>
            {indexes.length ? (
                <div>
                    <ListGroup
                        style={{
                            maxHeight: "165px",
                            overflowY: "auto",
                            marginBottom: "3px",
                            ...errorStyle
                        }}
                    >
                        {map(indexes, index => (
                            <ListGroupItem
                                key={index.id}
                                onClick={() => onSelect({ id: index.id, refId: index.reference.id })}
                            >
                                <Col xs={1}>
                                    <Checkbox checked={!!find(selected, ["id", index.id])} />
                                </Col>
                                <Col xs={8}>
                                    <strong>{index.reference.name}</strong>
                                </Col>
                                <Col xs={3}>
                                    Index Version <Label>{index.version}</Label>
                                </Col>
                            </ListGroupItem>
                        ))}
                    </ListGroup>
                    <Col xs={12}>
                        <div className="input-form-error">
                            <span className="input-error-message">{error}</span>
                        </div>
                    </Col>
                </div>
            ) : (
                <NoneFound noun="source references" />
            )}
        </div>
    );
};

const MultiSummary = ({ algorithm, samples, selected }) => {
    let prefix;
    let suffix;

    if (selected.length) {
        prefix = `Start ${selected.length} ${getTaskDisplayName(algorithm)} ${selected.length > 1 ? "jobs" : "job"}`;

        if (samples) {
            suffix = ` ${this.props.samples.length} ${this.props.samples.length > 1 ? "samples" : "sample"}.`;
        }
    }

    return (
        <div style={{ float: "left" }}>
            {prefix}
            {suffix}
        </div>
    );
};

const getInitialState = () => ({
    algorithm: "pathoscope_bowtie",
    selected: [],
    error: ""
});

export class CreateAnalysis extends React.Component {
    constructor(props) {
        super(props);
        this.state = getInitialState();
    }

    handleSelectIndex = index => {
        this.setState({ selected: xorBy(this.state.selected, [index], "id"), error: "" });
    };

    handleSubmit = e => {
        e.preventDefault();

        if (!this.state.selected.length) {
            return this.setState({ error: "Please select reference(s)" });
        }

        this.props.onAnalyze(this.props.id, this.state.selected, this.state.algorithm, this.props.userId);
        this.props.onHide();
    };

    render() {
        const { selected, algorithm } = this.state;

        return (
            <Modal show={this.props.show} onHide={this.props.onHide} onExited={() => this.setState(getInitialState())}>
                <Modal.Header>New Analysis</Modal.Header>
                <form onSubmit={this.handleSubmit}>
                    <Modal.Body>
                        <AlgorithmSelect
                            value={algorithm}
                            onChange={e => this.setState({ algorithm: e.target.value })}
                            hasHmm={this.props.hasHmm}
                        />
                        <IndexSelect
                            indexes={this.props.indexes}
                            onSelect={this.handleSelectIndex}
                            selected={selected}
                            error={this.state.error}
                        />
                    </Modal.Body>
                    <Modal.Footer>
                        <MultiSummary
                            algorithm={algorithm}
                            samples={this.props.samples}
                            selected={this.state.selected}
                        />
                        <Button type="submit" bsStyle="primary" icon="play" disabled={!!this.state.error}>
                            Start
                        </Button>
                    </Modal.Footer>
                </form>
            </Modal>
        );
    }
}

const mapStateToProps = state => ({
    show: routerLocationHasState(state, "createAnalysis"),
    sampleId: get(state, "samples.detail.id"),
    indexes: state.analyses.readyIndexes
});

const mapDispatchToProps = dispatch => ({
    onAnalyze: (sampleId, references, algorithm, userId) => {
        forEach(references, entry => dispatch(analyze(sampleId, entry.refId, algorithm, userId)));
    },
    onHide: () => {
        dispatch(push({ ...window.location, state: { createAnalysis: false } }));
    }
});

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(CreateAnalysis);
