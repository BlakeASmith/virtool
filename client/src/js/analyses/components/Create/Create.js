import { forEach, xorBy } from "lodash-es";
import React from "react";
import { Modal } from "react-bootstrap";
import { connect } from "react-redux";
import { pushState } from "../../../app/actions";
import { AlgorithmSelect, Button, Flex, FlexItem } from "../../../base";
import { getSelectedDocuments } from "../../../samples/selectors";
import { analyze } from "../../actions";
import { IndexSelector } from "./IndexSelector";
import { SelectedSamples } from "./SelectedSamples";

const MultiSummary = ({ samples, selected }) => {
    const product = selected.length * samples.length;

    if (product === 0) {
        return null;
    }

    return (
        <div style={{ textAlign: "left" }}>
            {product} job
            {product === 1 ? "" : "s"} will be started
        </div>
    );
};

const getInitialState = ({ defaultSubtraction }) => ({
    algorithm: "pathoscope_bowtie",
    selected: [],
    subtraction: defaultSubtraction,
    error: ""
});

export class CreateAnalysis extends React.Component {
    constructor(props) {
        super(props);
        this.state = getInitialState(props);
    }

    handleEnter = () => {
        this.setState(getInitialState(this.props));
    };

    handleSelectAlgorithm = e => {
        this.setState({ algorithm: e.target.value });
    };

    handleSelectIndex = index => {
        this.setState({ selected: xorBy(this.state.selected, [index], "id"), error: "" });
    };

    handleSelectSubtraction = e => {
        this.setState({ subtraction: e.target.value });
    };

    handleSubmit = e => {
        e.preventDefault();

        if (!this.state.selected.length) {
            return this.setState({ error: "Please select reference(s)" });
        }

        this.props.onAnalyze(
            this.props.documents,
            this.state.selected,
            this.state.algorithm,
            this.state.subtraction,
            this.props.userId
        );

        this.props.onHide();
    };

    render() {
        const { selected, algorithm } = this.state;

        const show = !!(this.props.documents && this.props.documents.length);

        return (
            <Modal show={show} onHide={this.props.onHide} onEnter={this.handleEnter}>
                <Modal.Header>Analyze</Modal.Header>
                <form onSubmit={this.handleSubmit}>
                    <Modal.Body>
                        <SelectedSamples samples={this.props.documents} />
                        <AlgorithmSelect
                            value={algorithm}
                            onChange={this.handleSelectAlgorithm}
                            hasHmm={this.props.hasHmm}
                        />
                        <IndexSelector
                            indexes={this.props.indexes}
                            onSelect={this.handleSelectIndex}
                            selected={selected}
                            error={this.state.error}
                        />
                        <Subtrac
                    </Modal.Body>
                    <Modal.Footer>
                        <Flex alignItems="center">
                            <FlexItem grow={1}>
                                <MultiSummary
                                    algorithm={algorithm}
                                    samples={this.props.documents}
                                    selected={this.state.selected}
                                />
                            </FlexItem>
                            <FlexItem>
                                <Button type="submit" bsStyle="primary" icon="play" disabled={!!this.state.error}>
                                    Start
                                </Button>
                            </FlexItem>
                        </Flex>
                    </Modal.Footer>
                </form>
            </Modal>
        );
    }
}

const mapStateToProps = state => ({
    documents: getSelectedDocuments(state),
    hasHmm: !!state.hmms.total_count,
    indexes: state.analyses.readyIndexes,
    defaultSubtraction: state.samples.detail.subtraction.id,
    userId: state.account.id
});

const mapDispatchToProps = dispatch => ({
    onAnalyze: (samples, references, algorithm, subtractionId, userId) => {
        forEach(samples, ({ id }) => {
            forEach(references, ({ refId }) => {
                dispatch(analyze(id, refId, algorithm, subtractionId, userId));
            });
        });
    },
    onHide: () => {
        dispatch(pushState({}));
    }
});

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(CreateAnalysis);
