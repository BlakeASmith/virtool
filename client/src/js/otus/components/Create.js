import { get } from "lodash-es";
import React from "react";
import { Modal } from "react-bootstrap";
import { connect } from "react-redux";
import { pushState } from "../../app/actions";
import { clearError } from "../../errors/actions";
import { getTargetChange, routerLocationHasState } from "../../utils/utils";
import { createOTU } from "../actions";
import { getNextState } from "../utils";
import OTUForm from "./OTUForm";

const getInitialState = () => ({
    name: "",
    abbreviation: "",
    errorName: "",
    errorAbbreviation: "",
    error: ""
});

class CreateOTU extends React.Component {
    constructor(props) {
        super(props);
        this.state = getInitialState();
    }

    static getDerivedStateFromProps(nextProps, prevState) {
        return getNextState(prevState.error, nextProps.error);
    }

    handleChange = e => {
        const { name, value, error } = getTargetChange(e.target);

        this.setState({ [name]: value, [error]: "" });

        if (this.props.error) {
            this.props.onClearError("CREATE_OTU_ERROR");
        }
    };

    handleHide = () => {
        this.props.onHide(this.props);
    };

    handleModalExited = () => {
        this.setState(getInitialState());

        if (this.props.error) {
            this.props.onClearError("CREATE_OTU_ERROR");
        }
    };

    handleSubmit = e => {
        e.preventDefault();

        if (!this.state.name) {
            return this.setState({
                errorName: "Required Field"
            });
        }

        if (!this.state.errorName || !this.state.errorAbbreviation) {
            this.props.onSubmit(this.props.refId, this.state.name, this.state.abbreviation);
        }
    };

    render() {
        return (
            <Modal show={this.props.show} onHide={this.handleHide} onExited={this.handleModalExited}>
                <Modal.Header onHide={this.props.onHide} closeButton>
                    Create OTU
                </Modal.Header>
                <OTUForm
                    name={this.state.name}
                    abbreviation={this.state.abbreviation}
                    handleSubmit={this.handleSubmit}
                    handleChange={this.handleChange}
                    errorName={this.state.errorName}
                    errorAbbreviation={this.state.errorAbbreviation}
                />
            </Modal>
        );
    }
}

const mapStateToProps = state => ({
    show: routerLocationHasState(state, "createOTU"),
    error: get(state, "errors.CREATE_OTU_ERROR.message", ""),
    pending: state.otus.createPending,
    refId: state.references.detail.id
});

const mapDispatchToProps = dispatch => ({
    onSubmit: (refId, name, abbreviation) => {
        dispatch(createOTU(refId, name, abbreviation));
    },

    onHide: () => {
        dispatch(pushState({ createOTU: false }));
    },

    onClearError: error => {
        dispatch(clearError(error));
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(CreateOTU);
