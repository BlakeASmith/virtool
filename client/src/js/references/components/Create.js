import React from "react";
import { connect } from "react-redux";
import { Alert, ButtonToolbar, Modal } from "react-bootstrap";
import { upperFirst } from "lodash-es";
import ReferenceForm from "./Form";
import { createReference } from "../actions";
import { clearError } from "../../errors/actions";
import { Button } from "../../base";

const getInitialState = () => ({
    name: "",
    description: "",
    dataType: "genome",
    organism: "",
    isPublic: false,
    errorName: "",
    errorDataType: ""
});

export class CreateReference extends React.Component {

    constructor (props) {
        super(props);
        this.state = getInitialState();
    }

    handleChange = (e) => {
        const { name, value } = e.target;

        const errorType = `error${upperFirst(e.target.name)}`;

        this.setState({
            [name]: value,
            [errorType]: ""
        });
    };

    handleSubmit = (e) => {
        e.preventDefault();

        if (!this.state.name.length) {
            this.setState({ errorName: "Required Field" });
        }

        if (!this.state.dataType.length) {
            this.setState({ errorDataType: "Required Field" });
        }

        if (this.state.name.length && this.state.dataType.length) {
            this.props.onSubmit(
                this.state.name,
                this.state.description,
                this.state.dataType,
                this.state.organism,
                this.state.isPublic
            );
        }

    };

    toggleCheck = () => {
        this.setState({ isPublic: !this.state.isPublic });
    };

    render () {

        return (
            <form onSubmit={this.handleSubmit}>
                <Modal.Body>
                    <Alert bsStyle="info">
                        <strong>
                            Create an empty reference.
                        </strong>
                    </Alert>
                    <ReferenceForm
                        state={this.state}
                        onChange={this.handleChange}
                        toggle={this.toggleCheck}
                    />
                </Modal.Body>

                <Modal.Footer>
                    <ButtonToolbar className="pull-right">
                        <Button type="submit" icon="save" bsStyle="primary">
                            Save
                        </Button>
                    </ButtonToolbar>
                </Modal.Footer>
            </form>
        );
    }
}

const mapDispatchToProps = dispatch => ({

    onSubmit: (name, description, dataType, organism, isPublic) => {
        dispatch(createReference(name, description, dataType, organism, isPublic));
    },

    onClearError: (error) => {
        dispatch(clearError(error));
    }

});

export default connect(null, mapDispatchToProps)(CreateReference);
