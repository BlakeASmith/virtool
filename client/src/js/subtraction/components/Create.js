import React from "react";
import { filter } from "lodash";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import { Row, Col, ListGroup, Modal } from "react-bootstrap";

import { findFiles } from "../../files/actions";
import { createSubtraction } from "../actions";
import { Button, Icon, Input, ListGroupItem, RelativeTime } from "../../base";

const getInitialState = () => ({
    subtractionId: "",
    fileId: ""
});

class CreateSubtraction extends React.Component {

    constructor (props) {
        super(props);
        this.state = getInitialState();
    }

    modalEnter = () => {
        this.props.onFindFiles();
    };

    modalExited = () => {
        this.setState(getInitialState());
    };

    handleSubmit = (e) => {
        e.preventDefault();
        this.props.onCreate(this.state);
    };

    render () {

        let fileComponents = filter(this.props.files, {type: "subtraction"}).map(file => {
            const active = file.id === this.state.fileId;

            return (
                <ListGroupItem
                    key={file.id}
                    active={active}
                    onClick={() => this.setState({fileId: active ? "" : file.id})}
                >
                    <Row>
                        <Col xs={7}>
                            <strong>{file.name}</strong>
                        </Col>
                        <Col xs={5}>
                            Uploaded <RelativeTime time={file.uploaded_at} /> by {file.user.id}
                        </Col>
                    </Row>
                </ListGroupItem>
            );
        });

        if (!fileComponents.length) {
            fileComponents = (
                <ListGroupItem className="text-center">
                    <Icon name="info" /> No files found. <Link to="/subtraction/files">Upload some</Link>.
                </ListGroupItem>
            );
        }

        return (
            <Modal bsSize="large" show={this.props.show} onHide={this.props.onHide} onEnter={this.modalEnter}
                onExited={this.modalExited}>

                <Modal.Header>
                    Create Subtraction
                </Modal.Header>

                <form onSubmit={this.handleSubmit}>
                    <Modal.Body>
                        <Input
                            type="text"
                            label="Unique Name"
                            value={this.state.subtractionId}
                            onChange={(e) => this.setState({subtractionId: e.target.value})}
                        />

                        <h5><strong>Files</strong></h5>
                        <ListGroup>
                            {fileComponents}
                        </ListGroup>
                    </Modal.Body>

                    <Modal.Footer className="modal-footer">
                        <Button
                            type="submit"
                            bsStyle="primary"
                            icon="play"
                            pullRight
                        >
                            Start
                        </Button>
                    </Modal.Footer>
                </form>
            </Modal>
        );
    }
}

const mapStateToProps = (state) => ({
    files: state.files.documents
});

const mapDispatchToProps = (dispatch) => ({

    onFindFiles: () => {
        dispatch(findFiles("subtraction"));
    },

    onCreate: ({ subtractionId, fileId }) => {
        dispatch(createSubtraction(subtractionId, fileId));
    }
});

const Container = connect(mapStateToProps, mapDispatchToProps)(CreateSubtraction);

export default Container;
