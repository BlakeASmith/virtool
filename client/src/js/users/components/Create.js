/**
 * @license
 * The MIT License (MIT)
 * Copyright 2015 Government of Canada
 *
 * @author
 * Ian Boyes
 *
 */

import React from "react";
import { connect } from "react-redux";
import { Row, Col, Modal, ButtonToolbar } from "react-bootstrap";

import { createUser } from "../actions";
import { Icon, Input, Checkbox, Button } from "../../base";

const getInitialState = () => ({
    username: "",
    password: "",
    confirm: "",
    forceReset: false
});

class CreateUser extends React.PureComponent {

    constructor (props) {
        super(props);
        this.state = getInitialState();
    }

    handleSubmit = (event) => {
        event.preventDefault();
        this.props.onCreate(this.state.username, this.state.password, this.state.forceReset);
    };

    render = () => (
        <Modal show={this.props.show} onHide={this.props.onHide} onEnter={this.modalEnter}>
            <Modal.Header onHide={this.props.onHide} closeButton>
                Create User
            </Modal.Header>
            <form onSubmit={this.handleSubmit}>
                <Modal.Body>
                    <Row>
                        <Col xs={12}>
                            <Input
                                type="text"
                                name="username"
                                label="Username"
                                value={this.state.username}
                                onChange={(e) => this.setState({username: e.target.value})}
                            />
                        </Col>
                    </Row>
                    <Row>
                        <Col xs={6}>
                            <Input
                                type="password"
                                name="password"
                                label="Password"
                                value={this.state.password}
                                onChange={(e) => this.setState({password: e.target.value})}
                            />
                        </Col>
                        <Col xs={6}>
                            <Input
                                type="password"
                                name="confirm"
                                label="Confirm"
                                value={this.state.confirm}
                                onChange={(e) => this.setState({confirm: e.target.value})}
                            />
                        </Col>
                    </Row>
                    <Row>
                        <Col xs={12}>
                            <Checkbox
                                label="Force user to reset password on login"
                                checked={this.state.forceReset}
                                onClick={() => this.setState({forceReset: !this.state.forceReset})}
                            />
                        </Col>
                    </Row>
                </Modal.Body>
                <Modal.Footer>
                    <ButtonToolbar className="pull-right">
                        <Button bsStyle="primary" type="submit">
                            <Icon name="floppy"/> Save
                        </Button>
                    </ButtonToolbar>
                </Modal.Footer>
            </form>
        </Modal>
    );
}

const mapStateToProps = (state) => {
    return {
        error: state.users.createError,
        pending: state.users.createPending
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        onCreate: (userId, password, forceReset) => {
            dispatch(createUser(userId, password, forceReset));
        }
    };
};

const Container = connect(mapStateToProps, mapDispatchToProps)(CreateUser);

export default Container;
