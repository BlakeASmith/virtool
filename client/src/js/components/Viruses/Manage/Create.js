/**
 * @license
 * The MIT License (MIT)
 * Copyright 2015 Government of Canada
 *
 * @author
 * Ian Boyes
 *
 * @exports AddVirus
 */

import React from "react";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";
import { Row, Col, Modal, Alert, ButtonToolbar } from "react-bootstrap";

import { Icon, Flex, FlexItem, Input, Button } from "virtool/js/components/Base";
import { createVirusSetName, createVirusSetAbbreviation } from "../../../actions/VirusActions";

/**
 * A form for adding a new virus, defining its name and abbreviation.
 */
class CreateVirus extends React.Component {

    constructor (props) {
        super(props);
    }

    static propTypes = {
        name: React.PropTypes.string,
        abbreviation: React.PropTypes.string,
        onSetName: React.PropTypes.func,
        onSetAbbreviation: React.PropTypes.func
    };

    componentDidMount () {
        this.inputNode.focus();
    }

    render () {

        /*
        // An alert that can be shown when an error occurs when submitting the form.
        let alert;

        if (this.state.error) {
            let message;

            // Show an error when no virus name is defined. Getting this error doesn't require a trip to the server.
            if (this.state.error.unnamed) {
                message = "A virus name must be provided";
            }

            // Show an error when the provided name or abbreviation is already in use in the database. These errors are
            // sent by the server.
            else {
                if (this.state.error.name) {
                    message = "Name is already in use";
                }

                if (this.state.error.abbreviation) {
                    message = "Abbreviation is already in use";
                }

                if (this.state.error.name && this.state.error.abbreviation) {
                    message = "Name and abbreviation are already in use";
                }
            }

            // Construct the alert component. The component will be null if no errors are defined.
            alert = (
                <Alert bsStyle="danger">
                    <Flex>
                        <FlexItem grow={0} shrink={0}>
                            <Icon name="warning" />
                        </FlexItem>
                        <FlexItem grow={1} shrink={0} pad>
                            {message}
                        </FlexItem>
                    </Flex>
                </Alert>
            );
        }

        */

        const inputProps = {
            type: "text",
            onChange: this.handleChange
        };

        return (
            <Modal show={true} onHide={this.props.onClose} onExited={this.modalExited}>
                <Modal.Header onHide={this.props.onClose} closeButton>
                    New Virus
                </Modal.Header>

                <form onSubmit={this.handleSubmit}>
                    <Modal.Body>
                        <Row>
                            <Col md={9}>
                                <Input
                                    {...inputProps}
                                    ref={(node) => this.inputNode = node}
                                    label="Name"
                                    value={this.props.name}
                                    onChange={(e) => {this.props.onSetName(e.target.value)}}
                                />
                            </Col>
                            <Col md={3}>
                                <Input
                                    {...inputProps}
                                    label="Abbreviation"
                                    value={this.props.abbreviation}
                                    onChange={(e) => {this.props.onSetAbbreviation(e.target.value)}}
                                />
                            </Col>
                        </Row>
                    </Modal.Body>

                    <Modal.Footer>
                        <ButtonToolbar className="pull-right">
                            <Button type="submit" bsStyle="primary">
                                <Icon name="floppy"/> Save
                            </Button>
                        </ButtonToolbar>
                    </Modal.Footer>
                </form>
            </Modal>
        );
    }

}

const mapStateToProps = (state) => {
    return {
        name: state.createVirus.name,
        abbreviation: state.createVirus.abbreviation
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        onSetName: (name) => {
            dispatch(createVirusSetName(name));
        },

        onSetAbbreviation: (abbreviation) => {
            dispatch(createVirusSetAbbreviation(abbreviation));
        }
    };
};

const CreateVirusContainer = withRouter(connect(
    mapStateToProps,
    mapDispatchToProps
)(CreateVirus));

export default CreateVirusContainer;

