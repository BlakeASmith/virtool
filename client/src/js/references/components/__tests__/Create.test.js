import { CreateReference, mapDispatchToProps } from "../Create";

describe("<CreateReference />", () => {
    const e = {
        preventDefault: jest.fn(),
        target: {
            name: "name",
            value: "foo",
            error: "error"
        }
    };
    const props = {
        onSubmit: jest.fn()
    };
    let state;
    beforeEach(() => {
        state = {
            name: [{ foo: "bar" }],
            dataType: [{ foo: "bar" }],
            description: "foo",
            organism: "bar",
            dataType: [{ foo: "bar" }],
            errorDataType: "",
            errorName: "",
            mode: "foo"
        };
    });

    it("should call handleChange() when ReferenceForm is changed", () => {
        const wrapper = shallow(<CreateReference {...props} />);
        wrapper.setState({
            ...state
        });
        wrapper.find("ReferenceForm").simulate("change", e);
        expect(wrapper.state()).toEqual({
            ...state,
            name: "foo"
        });
    });

    it("handleSubmit() should return errorName when [this.state.name.length = 0]", () => {
        const wrapper = shallow(<CreateReference {...props} />);
        wrapper.setState({
            ...state,
            name: []
        });

        wrapper.find("form").simulate("submit", e);
        expect(wrapper.state()).toEqual({
            ...state,
            errorName: "Required Field",
            name: []
        });
    });

    it("handleSubmit() should return errorName when [this.state.dataType.length = 0]", () => {
        const wrapper = shallow(<CreateReference {...props} />);
        wrapper.setState({
            ...state,
            dataType: []
        });
        wrapper.find("form").simulate("submit", e);
        expect(wrapper.state()).toEqual({
            ...state,
            dataType: [],
            errorDataType: "Required Field"
        });
    });

    it("handleSubmit() should return onSubmit when [this.state.name.length!=0] and [this.state.dataType.length!=0]", () => {
        const wrapper = shallow(<CreateReference {...props} />);
        wrapper.setState({
            ...state
        });
        wrapper.find("form").simulate("submit", e);
        expect(props.onSubmit).toHaveBeenCalledWith([{ foo: "bar" }], "foo", [{ foo: "bar" }], "bar");
    });
});

describe("mapDispatchToProps()", () => {
    const dispatch = jest.fn();
    const props = mapDispatchToProps(dispatch);
    it("should return onSubmit() in props", () => {
        props.onSubmit("foo", "bar", "fee", "baz");
        expect(dispatch).toHaveBeenCalledWith({
            name: "foo",
            description: "bar",
            dataType: "fee",
            organism: "baz",
            type: "CREATE_REFERENCE_REQUESTED"
        });
    });

    it("should return onClearError() in props", () => {
        props.onClearError("foo");
        expect(dispatch).toHaveBeenCalledWith({ error: "foo", type: "CLEAR_ERROR" });
    });
});
