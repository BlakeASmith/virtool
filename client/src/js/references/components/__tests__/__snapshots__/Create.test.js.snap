// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateReference /> should render 1`] = `
<Create__Container>
  <form
    onSubmit={[Function]}
  >
    <Alert>
      <strong>
        Create an empty reference.
      </strong>
    </Alert>
    <ReferenceForm
      description=""
      errorName=""
      mode="create"
      name=""
      onChange={[Function]}
      organism=""
    />
    <DataTypeSelection
      dataType="genome"
      onSelect={[Function]}
    />
    <Modal__DialogFooter>
      <ButtonToolbar>
        <Button
          bsStyle="primary"
          icon="save"
          pullRight={false}
          tipPlacement="top"
          type="submit"
        >
          Save
        </Button>
      </ButtonToolbar>
    </Modal__DialogFooter>
  </form>
</Create__Container>
`;
